File: VoidMail/Views/Inbox/InboxView.swift
import SwiftUI

struct InboxView: View {
    @StateObject private var gmailService = GmailService.shared
    @StateObject private var authService = GoogleAuthService.shared
    @EnvironmentObject var notificationManager: InAppNotificationManager
    @State private var selectedTab = 0
    @State private var selectedEmail: Email?
    @State private var searchText = ""
    @State private var showHelix = false
    @State private var selectedAccountEmail: String? = nil  // nil = All accounts
    @State private var readFilter: ReadFilter = .read

    enum ReadFilter: String, CaseIterable {
        case all = "All"
        case unread = "Unread"
        case read = "Read"

        var icon: String {
            switch self {
            case .all: return "envelope.fill"
            case .unread: return "envelope.badge.fill"
            case .read: return "envelope.open.fill"
            }
        }

        /// Label shown in the pill (no text for "All" state)
        var pillLabel: String {
            switch self {
            case .all: return ""
            case .unread: return "Unread"
            case .read: return "Read"
            }
        }
    }

    private let chips = ["All", "Priority", "Updates", "Newsletters"]

    private var unreadCount: Int {
        filteredEmails.filter { !$0.isRead }.count
    }

    private var primaryAccountColor: Color {
        if let selected = selectedAccountEmail,
           let account = authService.accounts.first(where: { $0.email == selected }) {
            return account.colorTag.color
        }
        return authService.accounts.first?.colorTag.color ?? .accentSkyBlue
    }

    /// Get account color for a specific email
    private func accountColor(for email: Email) -> Color {
        if let accountEmail = email.accountEmail,
           let account = authService.accounts.first(where: { $0.email == accountEmail }) {
            return account.colorTag.color
        }
        return primaryAccountColor
    }

    /// Display name for the currently selected account
    private var selectedAccountLabel: String {
        if let email = selectedAccountEmail {
            return authService.getAccountName(for: email)
        }
        return "All"
    }

    var body: some View {
        NavigationStack {
            ZStack(alignment: .bottomTrailing) {
                Color.bgDeep.ignoresSafeArea()

                VStack(spacing: 0) {
                    // MARK: Screen Header
                    VStack(alignment: .leading, spacing: 2) {
                        HStack {
                            Text("VOIDMAIL")
                                .metaLabel()
                            Spacer()
                            HStack(spacing: 8) {
                                Text("\(unreadCount) UNREAD")
                                    .font(Typo.mono)
                                    .tracking(1)
                                    .foregroundColor(unreadCount > 0 ? .accentYellow : .textTertiary)

                                // Refresh button
                                Button {
                                    Task { await gmailService.sync() }
                                } label: {
                                    ZStack {
                                        Circle()
                                            .fill(Color.accentGreen.opacity(0.12))
                                            .frame(width: 34, height: 34)
                                        if gmailService.isSyncing {
                                            ProgressView()
                                                .tint(.accentGreen)
                                                .scaleEffect(0.7)
                                        } else {
                                            Image(systemName: "arrow.clockwise")
                                                .font(.system(size: 15, weight: .semibold))
                                                .foregroundColor(.accentGreen)
                                        }
                                    }
                                }
                                .disabled(gmailService.isSyncing)

                                // AI Helix button
                                Button { showHelix = true } label: {
                                    ZStack {
                                        Circle()
                                            .fill(Color.accentSkyBlue.opacity(0.12))
                                            .frame(width: 34, height: 34)
                                        Image(systemName: "sparkles")
                                            .font(.system(size: 15, weight: .semibold))
                                            .foregroundColor(.accentSkyBlue)
                                    }
                                }
                            }
                        }

                        // Inbox title + inline filter pills (all on one row)
                        HStack(alignment: .firstTextBaseline, spacing: 8) {
                            Text("Inbox")
                                .font(.system(size: 48, weight: .heavy))
                                .foregroundColor(.textPrimary)
                                .textCase(.uppercase)
                                .tracking(-1.5)
                                .fixedSize()
                                .layoutPriority(1)

                            Spacer(minLength: 4)

                            // Account dropdown pill
                            Menu {
                                Button {
                                    withAnimation(.easeInOut(duration: 0.2)) { selectedAccountEmail = nil }
                                } label: {
                                    HStack {
                                        Label("All Inboxes", systemImage: "tray.2.fill")
                                        if selectedAccountEmail == nil { Image(systemName: "checkmark") }
                                    }
                                }
                                Divider()
                                ForEach(authService.accounts) { account in
                                    Button {
                                        withAnimation(.easeInOut(duration: 0.2)) { selectedAccountEmail = account.email }
                                    } label: {
                                        HStack {
                                            Label(account.label, systemImage: "envelope.fill")
                                            if selectedAccountEmail == account.email { Image(systemName: "checkmark") }
                                        }
                                    }
                                }
                            } label: {
                                HStack(spacing: 4) {
                                    if let email = selectedAccountEmail,
                                       let account = authService.accounts.first(where: { $0.email == email }) {
                                        Circle()
                                            .fill(account.colorTag.color)
                                            .frame(width: 6, height: 6)
                                    }
                                    Text(selectedAccountLabel)
                                        .font(Typo.meta)
                                        .tracking(0.5)
                                        .foregroundColor(.textPrimary)
                                        .lineLimit(1)
                                        .truncationMode(.tail)
                                        .frame(maxWidth: 80)
                                    Image(systemName: "chevron.down")
                                        .font(.system(size: 9, weight: .semibold))
                                        .foregroundColor(.textTertiary)
                                }
                                .padding(.horizontal, 10)
                                .padding(.vertical, 5)
                                .background(Color.bgCardHover)
                                .clipShape(Capsule())
                                .fixedSize()
                            }

                            // Read/Unread filter pill
                            Menu {
                                Button {
                                    withAnimation(.easeInOut(duration: 0.2)) { readFilter = .unread }
                                } label: {
                                    HStack {
                                        Label("Unread", systemImage: "envelope.badge.fill")
                                        if readFilter == .unread { Image(systemName: "checkmark") }
                                    }
                                }
                                Button {
                                    withAnimation(.easeInOut(duration: 0.2)) { readFilter = .read }
                                } label: {
                                    HStack {
                                        Label("Read", systemImage: "envelope.open.fill")
                                        if readFilter == .read { Image(systemName: "checkmark") }
                                    }
                                }
                                if readFilter != .all {
                                    Divider()
                                    Button(role: .destructive) {
                                        withAnimation(.easeInOut(duration: 0.2)) { readFilter = .all }
                                    } label: {
                                        Label("Clear Filter", systemImage: "xmark.circle")
                                    }
                                }
                            } label: {
                                HStack(spacing: 4) {
                                    Image(systemName: readFilter.icon)
                                        .font(.system(size: 11))
                                        .foregroundColor(readFilter == .all ? .textTertiary : .accentSkyBlue)
                                    Text(readFilter == .all ? "Filter" : readFilter.pillLabel)
                                        .font(Typo.meta)
                                        .tracking(0.5)
                                        .foregroundColor(readFilter == .all ? .textPrimary : .accentSkyBlue)
                                    Image(systemName: "chevron.down")
                                        .font(.system(size: 9, weight: .semibold))
                                        .foregroundColor(.textTertiary)
                                }
                                .padding(.horizontal, 10)
                                .padding(.vertical, 5)
                                .background(readFilter == .all ? Color.bgCardHover : Color.accentSkyBlue.opacity(0.12))
                                .clipShape(Capsule())
                                .fixedSize()
                            }
                        }
                        .padding(.top, -6)
                    }
                    .padding(.horizontal, 20)
                    .padding(.top, 8)

                    // MARK: Filter Chips
                    FilterChipBar(chips: chips, selected: $selectedTab)
                        .padding(.horizontal, 20)
                        .padding(.top, 12)
                        .padding(.bottom, 12)

                    // MARK: Email List
                    emailList
                }

                // FAB moved to ContentView to sit next to floating nav bar
            }
            .navigationBarHidden(true)
            .sheet(isPresented: $showHelix) {
                HelixO1View()
            }
            .navigationDestination(item: $selectedEmail) { email in
                EmailDetailView(email: email)
            }
        }
    }

    // MARK: - Email List

    private var emailList: some View {
        ScrollView {
            LazyVStack(spacing: 0) {
                if gmailService.isLoading && gmailService.emails.isEmpty {
                    // Shimmer skeleton while initial load
                    ForEach(0..<6, id: \.self) { _ in
                        EmailSkeletonRow()
                    }
                } else if filteredEmails.isEmpty {
                    EmptyStateView(
                        icon: readFilter == .unread ? "envelope.open" : "tray",
                        title: "No Emails",
                        subtitle: readFilter == .unread ? "You're all caught up" :
                                  readFilter == .read ? "No read emails" :
                                  selectedTab > 0 ? "Nothing in this category" : "Your inbox is empty"
                    )
                    .frame(maxWidth: .infinity)
                    .padding(.top, 60)
                } else {
                    let grouped = groupedEmails
                    let sortedKeys = grouped.keys.sorted { key1, key2 in
                        if key1 == "Today" { return true }
                        if key2 == "Today" { return false }
                        if key1 == "Yesterday" { return true }
                        if key2 == "Yesterday" { return false }
                        return key1 > key2
                    }

                    ForEach(sortedKeys, id: \.self) { dateKey in
                        // Date divider
                        DateDivider(label: dateKey)
                            .padding(.horizontal, 20)
                            .padding(.top, 20)
                            .padding(.bottom, 10)

                        // Email cards
                        ForEach(grouped[dateKey] ?? []) { email in
                            SwipeableEmailRow(
                                email: email,
                                accountColor: accountColor(for: email),
                                onTap: { selectedEmail = email },
                                onDelete: { Task { await gmailService.deleteEmail(email.id) } },
                                onToggleRead: { Task { await gmailService.toggleRead(email.id) } }
                            )
                        }
                    }
                }
            }
            .padding(.bottom, 140)
        }
        .refreshable {
            await gmailService.sync()
        }
        .task {
            // Wire in-app notifications
            gmailService.inAppNotificationManager = notificationManager
            // Auto-fetch emails on first appear
            if gmailService.emails.isEmpty {
                await gmailService.fetchEmails()
            }
            // Start auto-refresh timer
            gmailService.startAutoRefresh()
        }
    }

    // MARK: - Grouped Emails

    private var filteredEmails: [Email] {
        var result = gmailService.emails

        // Filter by account
        if let selectedAccount = selectedAccountEmail {
            result = result.filter { $0.accountEmail == selectedAccount }
        }

        // Filter by read/unread
        switch readFilter {
        case .unread:
            result = result.filter { !$0.isRead }
        case .read:
            result = result.filter { $0.isRead }
        case .all:
            break
        }

        // Filter by chip
        if selectedTab > 0 {
            let category = EmailCategory.allCases[selectedTab - 1]
            result = result.filter { $0.category == category }
        }

        if !searchText.isEmpty {
            result = result.filter {
                $0.subject.localizedCaseInsensitiveContains(searchText) ||
                $0.from.displayName.localizedCaseInsensitiveContains(searchText) ||
                $0.snippet.localizedCaseInsensitiveContains(searchText)
            }
        }
        return result
    }

    private var groupedEmails: [String: [Email]] {
        let calendar = Calendar.current
        var groups: [String: [Email]] = [:]
        for email in filteredEmails {
            let key: String
            if calendar.isDateInToday(email.date) {
                key = "Today"
            } else if calendar.isDateInYesterday(email.date) {
                key = "Yesterday"
            } else {
                let formatter = DateFormatter()
                formatter.dateStyle = .medium
                key = formatter.string(from: email.date)
            }
            groups[key, default: []].append(email)
        }
        return groups
    }
}

// MARK: - Skeleton Loading Row

private struct EmailSkeletonRow: View {
    var body: some View {
        HStack(alignment: .top, spacing: 14) {
            // Avatar placeholder
            RoundedRectangle(cornerRadius: 8)
                .fill(Color.clear)
                .frame(width: 44, height: 44)
                .overlay(ShimmerView().cornerRadius(8))

            VStack(alignment: .leading, spacing: 8) {
                // Sender + time
                HStack {
                    RoundedRectangle(cornerRadius: 4)
                        .fill(Color.clear)
                        .frame(width: 120, height: 14)
                        .overlay(ShimmerView().cornerRadius(4))
                    Spacer()
                    RoundedRectangle(cornerRadius: 4)
                        .fill(Color.clear)
                        .frame(width: 50, height: 12)
                        .overlay(ShimmerView().cornerRadius(4))
                }
                // Subject
                RoundedRectangle(cornerRadius: 4)
                    .fill(Color.clear)
                    .frame(height: 14)
                    .overlay(ShimmerView().cornerRadius(4))
                // Snippet
                RoundedRectangle(cornerRadius: 4)
                    .fill(Color.clear)
                    .frame(width: 200, height: 12)
                    .overlay(ShimmerView().cornerRadius(4))
            }
        }
        .padding(20)
        .background(Color.bgEmailRow)
        .cornerRadius(8)
        .padding(.horizontal, 16)
        .padding(.vertical, 4)
    }
}

// MARK: - Swipeable Email Row

private struct SwipeableEmailRow: View {
    let email: Email
    let accountColor: Color
    let onTap: () -> Void
    let onDelete: () -> Void
    let onToggleRead: () -> Void

    @State private var offset: CGFloat = 0
    @State private var isSwiped = false

    private let swipeThreshold: CGFloat = 80
    private let actionWidth: CGFloat = 80

    var body: some View {
        ZStack(alignment: .center) {
            // Background actions revealed on swipe
            HStack(spacing: 0) {
                // Left action (swipe right → toggle read)
                HStack {
                    Spacer()
                    VStack(spacing: 4) {
                        Image(systemName: email.isRead ? "envelope.badge.fill" : "envelope.open.fill")
                            .font(.system(size: 20, weight: .semibold))
                        Text(email.isRead ? "Unread" : "Read")
                            .font(.system(size: 11, weight: .bold, design: .monospaced))
                    }
                    .foregroundColor(.white)
                    .frame(width: actionWidth)
                }
                .frame(maxWidth: max(offset, 0), maxHeight: .infinity)
                .background(Color.accentSkyBlue)

                Spacer()

                // Right action (swipe left → delete)
                HStack {
                    VStack(spacing: 4) {
                        Image(systemName: "trash.fill")
                            .font(.system(size: 20, weight: .semibold))
                        Text("Delete")
                            .font(.system(size: 11, weight: .bold, design: .monospaced))
                    }
                    .foregroundColor(.white)
                    .frame(width: actionWidth)
                    Spacer()
                }
                .frame(maxWidth: max(-offset, 0), maxHeight: .infinity)
                .background(Color.accentPink)
            }
            .cornerRadius(8)
            .padding(.horizontal, 16)

            // Foreground email row
            EmailRowView(email: email, accountColor: accountColor)
                .offset(x: offset)
                .gesture(
                    DragGesture(minimumDistance: 20)
                        .onChanged { value in
                            // Only allow horizontal swipes
                            if abs(value.translation.width) > abs(value.translation.height) {
                                offset = value.translation.width
                            }
                        }
                        .onEnded { value in
                            let dragWidth = value.translation.width

                            if dragWidth > swipeThreshold {
                                // Swiped right → toggle read
                                triggerAction(isDelete: false)
                            } else if dragWidth < -swipeThreshold {
                                // Swiped left → delete
                                triggerAction(isDelete: true)
                            } else {
                                // Snap back
                                withAnimation(.spring(response: 0.3, dampingFraction: 0.7)) {
                                    offset = 0
                                }
                            }
                        }
                )
                .contentShape(Rectangle())
                .onTapGesture {
                    if offset == 0 {
                        onTap()
                    } else {
                        withAnimation(.spring(response: 0.3, dampingFraction: 0.7)) {
                            offset = 0
                        }
                    }
                }
        }
    }

    private func triggerAction(isDelete: Bool) {
        let targetOffset: CGFloat = isDelete ? -UIScreen.main.bounds.width : UIScreen.main.bounds.width

        withAnimation(.easeIn(duration: 0.2)) {
            offset = targetOffset
        }

        DispatchQueue.main.asyncAfter(deadline: .now() + 0.25) {
            if isDelete {
                onDelete()
            } else {
                onToggleRead()
            }
            // Reset after action
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
                offset = 0
            }
        }
    }
}

File: lib/views/inbox/inbox_view.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import '../../design_system/colors.dart';
import '../../design_system/typography.dart';
import '../../design_system/components.dart';
import '../../models/email.dart';
import '../../services/gmail_service.dart';
import '../../services/auth_service.dart';
import 'email_row_view.dart';
import 'email_detail_view.dart';

/// Main Inbox Interface with category filters, swipe actions, and date grouping
class InboxView extends StatefulWidget {
  final VoidCallback onHelixTap;

  const InboxView({super.key, required this.onHelixTap});

  @override
  State<InboxView> createState() => _InboxViewState();
}

class _InboxViewState extends State<InboxView> {
  String _selectedCategory = 'All';
  String? _selectedAccount;
  bool _showUnreadOnly = false;

  final _categories = ['All', 'Priority', 'Updates', 'Newsletters'];

  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance.addPostFrameCallback((_) {
      final gmail = context.read<GmailService>();
      if (gmail.emails.isEmpty) {
        gmail.loadMockEmails();
      }
    });
  }

  List<Email> _getFilteredEmails(GmailService gmail) {
    var emails = gmail.emails;

    // Filter by category
    if (_selectedCategory != 'All') {
      final category = EmailCategory.values.firstWhere(
        (c) => c.label == _selectedCategory,
        orElse: () => EmailCategory.primary,
      );
      emails = emails.where((e) => e.category == category).toList();
    }

    // Filter by account
    if (_selectedAccount != null) {
      emails =
          emails.where((e) => e.accountEmail == _selectedAccount).toList();
    }

    // Filter by unread
    if (_showUnreadOnly) {
      emails = emails.where((e) => !e.isRead).toList();
    }

    return emails;
  }

  Map<String, List<Email>> _groupByDate(List<Email> emails) {
    final Map<String, List<Email>> grouped = {};
    for (final email in emails) {
      final key = email.dateGroupKey;
      grouped.putIfAbsent(key, () => []).add(email);
    }
    return grouped;
  }

  @override
  Widget build(BuildContext context) {
    final gmail = context.watch<GmailService>();
    final auth = context.watch<AuthService>();
    final filtered = _getFilteredEmails(gmail);
    final grouped = _groupByDate(filtered);

    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        // Header
        _buildHeader(gmail, auth),

        // Category chips
        FilterChipBar(
          filters: _categories,
          selected: _selectedCategory,
          onSelected: (cat) => setState(() => _selectedCategory = cat),
        ),

        const SizedBox(height: 8),

        // Email list
        Expanded(
          child: gmail.isLoading
              ? _buildSkeletonList()
              : filtered.isEmpty
                  ? const EmptyStateView(
                      icon: Icons.inbox,
                      title: 'All clear',
                      subtitle: 'No emails match your filters',
                    )
                  : RefreshIndicator(
                      color: VoidColors.accentGreen,
                      backgroundColor: VoidColors.bgSurface,
                      onRefresh: () => gmail.fetchEmails(),
                      child: _buildEmailList(grouped, gmail),
                    ),
        ),
      ],
    );
  }

  Widget _buildHeader(GmailService gmail, AuthService auth) {
    return Padding(
      padding: const EdgeInsets.fromLTRB(20, 8, 20, 12),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          // Meta row
          Row(
            children: [
              Text(
                'VOIDMAIL',
                style: Typo.metaLabel,
              ),
              const Spacer(),
              // Unread count
              Text(
                '${gmail.unreadCount}',
                style: Typo.mono.copyWith(
                  color: gmail.unreadCount > 0
                      ? VoidColors.accentYellow
                      : VoidColors.textTertiary,
                  letterSpacing: 1,
                ),
              ),
              const SizedBox(width: 12),
              // Sync circle button
              GestureDetector(
                onTap: () => gmail.fetchEmails(),
                child: Container(
                  width: 34,
                  height: 34,
                  decoration: BoxDecoration(
                    color: VoidColors.accentGreen.withValues(alpha: 0.12),
                    shape: BoxShape.circle,
                  ),
                  child: Center(
                    child: gmail.isSyncing
                        ? const SizedBox(
                            width: 16,
                            height: 16,
                            child: CircularProgressIndicator(
                              strokeWidth: 2,
                              color: VoidColors.accentGreen,
                            ),
                          )
                        : const Icon(
                            Icons.refresh,
                            size: 16,
                            color: VoidColors.accentGreen,
                          ),
                  ),
                ),
              ),
              const SizedBox(width: 8),
              // Helix AI circle button
              GestureDetector(
                onTap: widget.onHelixTap,
                child: Container(
                  width: 34,
                  height: 34,
                  decoration: BoxDecoration(
                    color: VoidColors.accentSkyBlue.withValues(alpha: 0.12),
                    shape: BoxShape.circle,
                  ),
                  child: const Center(
                    child: Icon(
                      Icons.auto_awesome,
                      size: 16,
                      color: VoidColors.accentSkyBlue,
                    ),
                  ),
                ),
              ),
            ],
          ),
          const SizedBox(height: 4),
          // Title row with inline account dropdown + read filter
          Row(
            crossAxisAlignment: CrossAxisAlignment.end,
            children: [
              Text(
                'INBOX',
                style: Typo.inboxTitle,
              ),
              const SizedBox(width: 12),
              if (auth.accounts.length > 1)
                Padding(
                  padding: const EdgeInsets.only(bottom: 8),
                  child: TagChip(
                    label: _selectedAccount ?? 'All',
                    onTap: () {
                      _showAccountPicker(auth);
                    },
                  ),
                ),
              Padding(
                padding: const EdgeInsets.only(bottom: 8),
                child: TagChip(
                  label: _showUnreadOnly ? 'Unread' : 'All',
                  isActive: _showUnreadOnly,
                  onTap: () =>
                      setState(() => _showUnreadOnly = !_showUnreadOnly),
                ),
              ),
            ],
          ),
        ],
      ),
    );
  }

  void _showAccountPicker(AuthService auth) {
    showModalBottomSheet(
      context: context,
      backgroundColor: VoidColors.bgCard,
      builder: (context) => SafeArea(
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            ListTile(
              title: Text('All Accounts', style: Typo.body),
              onTap: () {
                setState(() => _selectedAccount = null);
                Navigator.pop(context);
              },
            ),
            ...auth.accounts.map((account) => ListTile(
                  title: Text(account.label, style: Typo.body),
                  subtitle: Text(account.email, style: Typo.monoSmall),
                  onTap: () {
                    setState(() => _selectedAccount = account.email);
                    Navigator.pop(context);
                  },
                )),
          ],
        ),
      ),
    );
  }

  Widget _buildSkeletonList() {
    return ListView.builder(
      itemCount: 6,
      itemBuilder: (context, index) => const ShimmerEmailRow(),
    );
  }

  Widget _buildEmailList(
      Map<String, List<Email>> grouped, GmailService gmail) {
    final groups = grouped.entries.toList();

    return ListView.builder(
      padding: const EdgeInsets.only(bottom: 100),
      itemCount: groups.fold<int>(0, (sum, g) => sum + g.value.length + 1),
      itemBuilder: (context, index) {
        int current = 0;
        for (final group in groups) {
          // Date divider
          if (index == current) {
            return DateDivider(label: group.key);
          }
          current++;

          // Emails in group
          for (int i = 0; i < group.value.length; i++) {
            if (index == current) {
              final email = group.value[i];
              return Column(
                children: [
                  EmailRowView(
                    email: email,
                    index: i,
                    onTap: () => _navigateToDetail(email),
                    onSwipeLeft: () => gmail.deleteEmail(email.id),
                    onSwipeRight: () => gmail.toggleRead(email.id),
                  ),
                  if (i < group.value.length - 1)
                    const Padding(
                      padding: EdgeInsets.only(left: 72),
                      child: Divider(
                        height: 0.5,
                        color: VoidColors.border,
                      ),
                    ),
                ],
              );
            }
            current++;
          }
        }
        return const SizedBox.shrink();
      },
    );
  }

  void _navigateToDetail(Email email) {
    // Mark as read
    if (!email.isRead) {
      context.read<GmailService>().toggleRead(email.id);
    }

    Navigator.of(context).push(
      PageRouteBuilder(
        pageBuilder: (context, animation, secondaryAnimation) =>
            EmailDetailView(email: email),
        transitionsBuilder: (context, animation, secondaryAnimation, child) {
          return FadeTransition(
            opacity: animation,
            child: SlideTransition(
              position: Tween<Offset>(
                begin: const Offset(0.05, 0),
                end: Offset.zero,
              ).animate(CurvedAnimation(
                parent: animation,
                curve: Curves.easeOutCubic,
              )),
              child: child,
            ),
          );
        },
        transitionDuration: const Duration(milliseconds: 350),
      ),
    );
  }
}
